### ISSUE TITLE
enhance types of mergeProps

### ISSUE THREAD

**Issue by bigmistqke** [solidjs/solid#1526]
### Describe the bug

`mergeProps` is a powerful tool, but the typing is a bit clunky.

### Your Example Website or App

https://playground.solidjs.com/anonymous/f901c3ec-82cb-4fbf-bc81-6e6583c67b64

### Steps to Reproduce the Bug or Issue

``` tsx
type Vector = [number, number];
type Props = { tuple: Vector };

const App = (props: Partial<Props>) => {
  const merged = mergeProps({ tuple: [1, 1] }, props);
  merged.tuple; // type is (number[] | Vector) & (number[] | Vector | undefined)
  // ideally types would not become less narrow by using mergeProps

  // typecasting works, but is clunky and not typesafe
  const merged2 = mergeProps({ tuple: [1, 1] as Vector }, props);
  merged2.tuple; // type is Vector
  const merged3 = mergeProps({ tuple: [1, 1] as Vector }, props) as Props;
  merged3.tuple; // type is Vector

  // no type-error
  const merged4 = mergeProps({ tuple: [1, 1, 1] }, props)
  // no type-error
  const merged5 = mergeProps({ tuple: [1, 1, 1] }, props) as Props

  return <></>;
};
```

### Expected behavior

1. ideally types would not become less narrow by using mergeProps and no additional typecasts would be needed
2. original typing of props should be respected when merging props, or there is at least a typesafe way to enforce this. 

### Platform

- OS: [e.g. macOS, Windows, Linux]
- Browser: [e.g. Chrome, Safari, Firefox]
- Version: [e.g. 91.1]

**Comment by otonashixav** [comment]
1. I mean this isn't really something fixable, typescript is going to infer `[1, 1]` as `number[]` and not `[number, number]`, and I don't see any way to correctly type this without `[1, 1] as const` or `[1, 1] satisfies Vector`.
2. `mergeProps` doesn't know which of the arguments passed is the original type, and we it doesn't make assumptions about what the user wants. You could make a wrapper that is meant to merge props onto default props and have the latter be the "original" type.

**Comment by bigmistqke** [comment]
2. I was thinking something in the line of mergeProps taking a generic that is the 'correct' type and the result of mergeProps would be typechecked against this type. Then `mergeProps<{tuple: Tuple}>({ tuple: [1, 1, 1] }, props)` would error and you could do something like `(props: Partial<Props>) => {const merged = mergeProps<Props>({...assign all values...}, props) ...}`

**Comment by otonashixav** [comment]
Right now the generic is the entire list of args, which is necessary to preserve their order so that we can make sure that the args that come later override those which come earlier. So you can currently use `mergeProps<{ tuple: Tuple }[]>`.

It should be possible to add an overload which takes a non-array generic to catch this usage, after the current one, if necessary. I'm not sure if that would be an improvement though.

**Comment by xdev1** [comment]
Not really 100% an answer to the question, but I personally have always asked myself (I'm not an actual Solid user yet) why this general `mergeProps` function  is always used explicitly for the "default props use case" in Solid (see Solid Tutorial ["Props/Default Props"](https://www.solidjs.com/tutorial/props_defaults?solved)), instead of using a special additional function tailored for exact that "default props use case" and nothing else. I think the above snippet is such a "default props use case". It may have some advantages if the arguments were swapped, means `f(props, defaultProps)` instead of `f(defaultProps, props)`, as `defaultProps` may be a large object literal, and so this argument order may be better to read and will not produce additional lines with prettier - maybe just a matter of taste :wink:
Okay, long story short: Let's try to define a function `withDefaults(props, defaultProps)` that covers only this "default props use case", hoping proper typing becomes simpler then (like said, I know, that was not exactly the question, but anyway).
My little brain did not make it to find a proper type for this `withDefaults` function with arity 2, I always had the same problems that have already described in the other comments above.
But if you implement this `withDefaults` function in curried form, typing seems to become easier: [Demo](https://playground.solidjs.com/anonymous/8092f03d-2aaf-4e91-b73d-ffec5856468d)

Maybe currying could also help with typing a method `mergeProps2(a)(b)(c)`, but I have not tried that.

**Comment by bigmistqke** [comment]
> Okay, long story short: Let's try to define a function withDefaults(props, defaultProps) that covers only this "default props use case", hoping proper typing becomes simpler then (like said, I know, that was not exactly the question, but anyway).

Not exactly the question, but exactly the intent behind my question: a typesafe way of defining default props 👍

**Comment by otonashixav** [comment]
```ts
// typecheck defaultProps as Props by assigning it
const defaultProps: Props = { tuple: [1, 1] };
const mergedProps = mergeProps(defaultProps, props);
// or in TS 4.9 and later
const mergedProps = mergeProps({ tuple: [1, 1] } satisfies Props, props);
```

**Comment by bigmistqke** [comment]
This does give us typesafety while assigning, but the types of the result-value do not accurately portray this:

```
type Vector = [number, number];
type Color = string | { r: number; g: number; b: number };
type Props = { tuple: Vector; color: Color };

const App = (props: Partial<Props>) => {
  const defaultProps: Props = { tuple: [1, 1], color: "red" };
  const merged1 = mergeProps(defaultProps, props);
  // type merged1.color = Color & (Color | undefined);
  const merged2 = withDefaults(props)(defaultProps);
  // type merged2.color = Color & (Color | undefined);
  const merged3 = { ...props, ...defaultProps };
  // type merged3.color = Color;
  return <></>;
};
```

[playground](https://playground.solidjs.com/anonymous/0fcc83d7-5af6-4c2e-9959-ec8ccbe3ddc5)

**Comment by xdev1** [comment]
It's a bit of a problem (at least it's confusing) that the base types are defined like this:

```ts
type Props = { tuple: Vector };

const App = (props: Partial<Props>) => {
  ...
}
``` 

instead of defining them in the usual way:

```ts
type Props = { tuple?: Vector };

const App = (props: Props) => {
  ...
}
```

Because then it will turn out that using the following pattern

```ts 
  const defaultProps: Props = { tuple: [1, 1] };
```

would not be very useful pattern any longer, but using the following will always work:

```ts 
  const defaultProps = { tuple: [1, 1] } satisfies Props;
```

Would really help if Solid Playground was using  a  TS version >= 4.9 :wink:

Also @bigmistqke, in your latest demo you have changed the return type of the `withDefaults` function: It's `=> P & D`, not just `=> P`.

**Comment by bigmistqke** [comment]
Edited the typo 👍

With `satisfies` and not using `Partial`-typehelper, `withDefaults` seems to be working.
+1 on Solid Playground using TS version >= 4.9

> It's a bit of a problem (at least it's confusing) that the base types are defined like this:

I think it's pretty clean tbh. Sad that it has implications on the typings.

**Comment by ryansolid** [comment]
Is there any action to take here?

**Comment by thetarnav** [comment]
You can contribute it to solid-primitives if you think it's useful 😉

**Comment by bigmistqke** [comment]
true, but I think in this case it would be good to have an 'official' solution that can be recommended in docs, since it's such a common task.

**Comment by ryansolid** [comment]
Ok, so basically TS can't do what we want, there is a solution with a different sort of API that uses currying which may be weird. I think the idea is interesting, but I'm also slow on adding new APIs to do something we already do a different way. Let me add this to my considerations for 2.0 list and in the meanwhile submit to solid-primitives to see how people like using it in practice.

**Comment by xdev1** [comment]
In the upcoming TS Version 5 there will be this thing called ["const type parameters"](https://devblogs.microsoft.com/typescript/announcing-typescript-5-0-beta/#const-type-parameters).

I currently have time to check that out the next few days, but maybe that feature could help with the "properly typed default props function" problem. Maybe I'm totally wrong and this hint is stupid, but I thing at least this sounds promising.

**Comment by otonashixav** [comment]
It wouldn't give you `Props` as the output type but it would work similarly to `as const` does now, but in the function declaration instead. Which would allow you to use an object literal in there without manually adding `as const` or `satisfies`, but similar to `as const`, comes at the cost of having (some) props become deeply readonly. Which has implications for typing, since types in typescript are rarely used in a const-correct manner. An example: https://playground.solidjs.com/anonymous/00355c0b-05ae-4257-8386-1f8f4207f9a0.

> This does give us typesafety while assigning, but the types of the result-value do not accurately portray this:
> I could cast of course

You don't need to cast, annotate the merged type with `Props` if you prefer that over the inferred intersections. That's still more work than typescript inferring the correct types, but at least you're not casting.

**Comment by lydianlights** [comment]
For what it's worth I just ran into this exact problem with `mergeProps` and I wrote my own helper function to resolve it:
```
import { type MergeProps, mergeProps } from "solid-js";

export function defaultProps<T, K extends keyof T>(
    defaults: Required<Pick<T, K>>,
    props: T
): MergeProps<[Required<Pick<T, K>>, T]> {
    const resolvedProps = mergeProps(defaults, props);
    return resolvedProps;
}
```

It doesn't merge arbitrary keys, it requires the merged keys to already exist on T (the ComponentProps). Additionally, it prevents types from "widening" (e.g. `"start" | "center" | "end"` does not get transformed into `string` in the final props). This is behavior that I would expect from default props. Essentially I made this helper to avoid having to explicitly constrain mergeProps every time I used it. I was actually going to submit an issue and a PR, but I see there's already been some discussion about this.

**Comment by xdev1** [comment]
@lydianlights Your solution is amazing! 
I'm wondering how much this can be simplified.

Do you think this may be a working simplification of your suggestion (obviously I've switched the two arguments):
[Demo](https://www.typescriptlang.org/play?#code/C4TwDgpgBAyglgWzAGzgMxAHgCoD4oC8U2UEAHsBAHYAmAzlAIZUhQD8UA3lANoDSUOFSgBrCCAD2aYgF0AXMX4yoAXygLsAbgBQ2tAFcqAY2BwJwmhDSN9yYAAUAThLB0cAGigDylWgzGS0ngAFNpQUGDOrhruYVCW1rbAdAoAShAAjvpwjhA0mPZwRiIeXri42gCUCvBIqBg4UABkUOlZOXkFRSXYnnzl+JxxwAAWzgDuUABEo9CIKBAI1MCMpuaCDFQSwIJIEo4rVDsjELlTOiq6APRXhARE5Ix10Pf3uqCQUABqECb7AEwAEUIvCo+gQACNTp4wZDTjIdB9oD8-o4AMzAog8WFQxww8G4-FwxwI97gaBOFwMIhDcIAN1+wABgLYChRTMcQJ09MZ+wxrO+vPRgO5TAFONOoohCjowEcQgA5hcdNojOZZVAAIJgMAg4KRKkKSmuSqEQZxNVUDW6ogJGx2Y10fVROieWnhKAM1FAhQ8ACMnj9MliHs9QoxvoDUCjQZDqkqKvClroEmQEAAdMgJAr9QmoDcoJgiwBaUtl0tQewAGQAopqYDWoABhAASNabAmwbeIAE17I2APIAMSgAAMwKOoG3Uo3y3P50XMLpwrlgPpHMIpjAJEsoBJ9MAwAf0yfztoVDogA)

... and if yes (if no, then please ignore the following) could this be an enhancement of your suggestion that does only allow optional props to be defaulted?

**Comment by bigmistqke** [comment]
@xdev1 I think it's probably best to keep it with the `MergeProps`-type so you don't have to cast unnecessarily, but I like your addition to only allow default values for optional types, that's pretty sweet.

**Comment by oddcelot** [comment]
> The signature of the resulting type is more narrow with this. I'm not too happy about the awkwardness of putting the "real" props as first argument though, to avoid running into typing issues with the spread operator 🫠 I have the feeling that I'm missing something…

Yeah, totally missed something 😄 
I case anyone still follows this issue: this [updated playground](https://playground.solidjs.com/anonymous/847a3e08-00f8-4b9b-9003-f937ff362ce2) may be interesting
